codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -428,6 +439,20 @@ public boolean merge(ChannelHandlerContext ctx, Http2RemoteFlowController.FlowCo
         }
     }
 
+    private void notifyLifecycleManagerOnError(ChannelFuture future, final ChannelHandlerContext ctx) {
+        if (!future.isDone()) {
```

Comment:
I'd feel better if this was an `assert !future.isDone()`, as if the future is done this method is being misused (`lifecycleManager.onError()` won't be called). But I guess that isn't strictly thread-safe :(, even though practically it probably would be safe.

Reply:
Actually no... its completely fine if `future.isDone()` is true  and this is called. I should just not be `successful` ;)

Reply:
How about you just remove this if then (add the listener unconditionally)? That is fully safe and only adds work if there was a race.

Reply:
@ejona86 maybe we misunderstand each other... so here is the calling code:

```java
            if (failureCause == null) {
                // This is best-effort as the operation is async it may still fail later on. We just update it here
                // directly as otherwise it may be challenging for the end-user to co-ordinate writes that depend
                // on the state.
                stream.pushPromiseSent();

                notifyLifecycleManagerOnError(future, ctx);
            } else {
                lifecycleManager.onError(ctx, failureCause);
            }
```

So if I remove the `!isDone()` check it will always add the listener even if the future was done but successfully. I wanted to avoid this...

Reply:
Ah, you're right. Then maybe the condition should be `if (!(future.isDone() && future.cause() == null))` as the check instead? I'd also accept renaming the method.

Right now there's tight coupling that's not easily visible between the caller and callee; you have to read both together to guarantee correctness. I want it to be more obvious "you are holding it wrong" (via rename) or for it to become fool-proof (the fully-precise check).